test/benchmark_http_client_test.cc

```diff
@@ -46,41 +67,52 @@ class BenchmarkClientHttpTest : public Test {
 
     auto& tracer = static_cast<Envoy::Tracing::MockHttpTracer&>(*http_tracer_);
     EXPECT_CALL(tracer, startSpan_(_, _, _, _))
-        .WillRepeatedly(
-            Invoke([&](const Envoy::Tracing::Config& config, const Envoy::Http::HeaderMap&,
-                       const Envoy::StreamInfo::StreamInfo&,
-                       const Envoy::Tracing::Decision) -> Envoy::Tracing::Span* {
-              EXPECT_EQ(Envoy::Tracing::OperationName::Egress, config.operationName());
-              auto* span = new NiceMock<Envoy::Tracing::MockSpan>();
-              return span;
-            }));
-    request_generator_ = []() {
-      auto header = std::make_shared<Envoy::Http::TestRequestHeaderMapImpl>(
-          std::initializer_list<std::pair<std::string, std::string>>(
-              {{":scheme", "http"}, {":method", "GET"}, {":path", "/"}, {":host", "localhost"}}));
-      return std::make_unique<RequestImpl>(header);
-    };
+        .WillRepeatedly([&](const Envoy::Tracing::Config& config, const Envoy::Http::HeaderMap&,
+                            const Envoy::StreamInfo::StreamInfo&,
+                            const Envoy::Tracing::Decision) -> Envoy::Tracing::Span* {
+          EXPECT_EQ(Envoy::Tracing::OperationName::Egress, config.operationName());
+          auto* span = new NiceMock<Envoy::Tracing::MockSpan>();
+          return span;
+        });
   }
 
-  void testBasicFunctionality(const uint64_t max_pending, const uint64_t connection_limit,
-                              const uint64_t amount_of_request) {
+  // void setupBenchmarkClient(const RequestGenerator& request_generator) {
+  //   client_ = std::make_unique<Client::BenchmarkClientHttpImpl>(
+  //       *api_, *dispatcher_, store_, std::make_unique<StreamingStatistic>(),
+  //       std::make_unique<StreamingStatistic>(), std::make_unique<StreamingStatistic>(),
+  //       std::make_unique<StreamingStatistic>(), false, cluster_manager_, http_tracer_,
+  //       "benchmark", request_generator, true);
+  // }
+  // Primary testing method. Confirms that connection limits are met and number of requests are
+  // correct. If not empty, also checks the header expectations, if empty, it is ignored.
+  void TestBenchmarkClientProcessesExpectedInflightRequests(
+      const uint64_t max_pending, const uint64_t connection_limit, const uint64_t amount_of_request,
+      const RequestGenerator& request_generator,
+      const std::vector<absl::flat_hash_set<std::string>>* header_expectations = nullptr) {
```

Comment:
I usually prefer using absl::Optional for pointers so the caller knows that a null is acceptable.

Reply:
I think that optional is not recommended for this, and a const pointer is preferred. This is to avoid making unnecessary copies.